src/advanced.sml

```diff
@@ -0,0 +1,416 @@
+  end
+
+  (* Returns whether a card is definitely playable *)
+  fun isDefinitelyPlayable (s : state) (is : info list) : bool =
+  let
+    val sus = SSet.toList (possibleSuits is)
+    val rs = RSet.toList (possibleRanks is)
+  in
+    length rs = 1 andalso List.all (fn su => hd rs = SD.lookup (#inPlay s) su + 1) sus
+  end
+
+  datatype fuzzyBool = True | False | Maybe
+
+  fun isPlayableFuzzy (s : state) (is : info list) : fuzzyBool =
+  let
+    val (sus', rs') = possibleSuitsAndRanks is
+    val sus = SSet.toList sus'
+    val rs = RSet.toList rs'
+  in
+    if length rs = 1 andalso List.all (fn su => hd rs = SD.lookup (#inPlay s) su + 1) sus
+    then True
+    else if List.all (fn su => not (RSet.member rs' (SD.lookup (#inPlay s) su + 1))) sus
+    then False
+    else
+        Maybe
+  end
+
+  fun FuzzyBoolToBool (k : fuzzyBool) (default : bool) : bool =
+  case k of
+       True => true
+     | False => false
+     | Maybe => default
+
+  (* check whether a card in your own hand is definitely not vital *)
+  fun isNotVital (s : state) (is : info list) : bool =
+  let
+    val rs = possibleRanks is
+    val sus = possibleSuits is
+  in
+    if RSet.member rs 5 then false else
+      List.all (fn su => List.all (fn r => isUseless s (su, r) orelse not (isVital s (su, r)))
+				  (RSet.toList rs))
+	       (SSet.toList sus)
+
+  end
+
+
+  (* Very simple strategy:
+   *
+   * A number clue including the oldest unclued card is a save clue.
+   * Other clues are play clues on the newest clued card.
+   *
+   * Look at previous action. If it was a play clue for us, play it.
+   * Otherwise, look at next player's hand.
+   *   If a play clue can be given, give it.
+   *   Otherwise, if the oldest unclued card is vital, give a save clue.
+   *   Otherwise, discard oldest card known to be already played.
+   *   Otherwise, discard oldest unclued card
+   *   Otherwise, try to give to the next player which cannot be interpreted as a play hint
+   *   Otherwise, discard the least vital card
+   *   Otherwise you have 8 clues, and then you clue 1 to the next player.
+   *)
+
+  (* Is this action a number clue to me which applies to the oldest card
+   * about which we previously had no positive information?
+   * TODO:
+   * - if no unplayed card of that (non-5) number are discarded then it's not a save
+   * - if the clue pointed at 1 card and it's playable assuming it's not useless, then it's a play*)
+  fun isSaveClueToMe (s : state) ((_,HintedRank (Me,r,cs)) : player * play) =
+    (case oldestUnclued (map tl (#clues s)) of
+          NONE => false
+        | SOME i => (case hd (List.nth (#clues s,i)) of
+                          IsRank _ => true
+                        | _ => false))
+    | isSaveClueToMe s _ = false
+
+  (* Is a rank hint which hints the cards in positions l interpreted as a save hint? *)
+  fun isSaveClue (s : state) (pl : player) (l : int list) : bool =
```

Comment:
I removed it in `simple.sml`, where it wasn't used indeed. I do use it in `advanced.sml`.